src/ripple/shamap/impl/SHAMapSync.cpp

```diff
@@ -332,51 +336,56 @@ bool SHAMap::getNodeFat (SHAMapNodeID wanted, std::vector<SHAMapNodeID>& nodeIDs
         return false;
     }
 
-    int count;
-    bool skipNode = false;
-    do
-    {
-
-        if (skipNode)
-            skipNode = false;
-        else
-        {
-            Serializer s;
-            node->addRaw (s, snfWIRE);
-            nodeIDs.push_back (wanted);
-            rawNodes.push_back (std::move (s.peekData ()));
-        }
+    std::stack<std::tuple <SHAMapTreeNode*, SHAMapNodeID, int>> stack;
+    stack.emplace (node, nodeID, depth);
 
-        if ((!fatRoot && wanted.isRoot ()) || node->isLeaf ()) // don't get a fat root_, can't get a fat leaf
-            return true;
+    while (! stack.empty ())
+    {
+        std::tie (node, nodeID, depth) = stack.top ();
+        stack.pop ();
 
-        SHAMapTreeNode* nextNode = nullptr;
-        SHAMapNodeID nextNodeID;
+        // Add this node to the reply
+        Serializer s;
+        node->addRaw (s, snfWIRE);
+        nodeIDs.push_back (nodeID);
+        rawNodes.push_back (std::move (s.peekData()));
 
-        count = 0;
-        for (int i = 0; i < 16; ++i)
+        if (node->isInner())
         {
-            if (!node->isEmptyBranch (i))
+            // We descend inner nodes with only a single child
+            // without decrementing the depth
+            int bc = node->getBranchCount();
+            if ((depth > 0) || (bc == 1))
             {
-                SHAMapNodeID nextNodeID = wanted.getChildNodeID (i);
-                nextNode = descendThrow (node, i);
-                ++count;
-                if (fatLeaves || nextNode->isInner ())
+                // We need to process this node's children
+                for (int i = 0; i < 16; ++i)
```

Comment:
The question "Why 16" could be addressed in the comment.  Something like, `Each node has exactly 16 children.  Process them all.`

To me, naming `16` locally locally has the same value as adding the information in the comment.  If there were a namespace scoped named constant, say, `shamapChildNodes` then we should probably use that constant.  We're writing code not only for ourselves but for auditors as well.